File: src/main/java/ex4/Exercise4.java
package ex4;

import util.FileReader;

import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.function.Predicate;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class Exercise4 {
    public static void main(String[] args) {

        part2();

    }

    private static void part2() {
        System.out.println(readInput()
                .map(list -> list.stream()
                        .map(str -> Arrays.stream(str.split("")).sorted().collect(Collectors.joining()))
                        .collect(Collectors.toList()))
                .filter(hasNoDuplicates)
                .count());
    }

    private static void part1() {
        System.out.println(readInput()
                .filter(hasNoDuplicates)
                .count());
    }

    private static Predicate<List<String>> hasNoDuplicates = l -> l.size() == new HashSet<>(l).size();

    private static Stream<List<String>> readInput() {
        return FileReader.readLines("input.txt", Exercise4.class, "\\s+").stream();
    }
}

File: src/main/kotlin/ex4/Exercise4.kt
package ex4


fun main(args: Array<String>) {
    part1()
    part2()
}

private fun part2() {
    println(FileReader.readLines("ex4/input.txt", " ")
            .map { it.map { it.split("").sorted().joinToString("") } }
            .filter { it.distinct().size == it.size }
            .size)
}

private fun part1() {
    println(FileReader.readLines("ex4/input.txt", " ")
            .filter { it.distinct().size == it.size }
            .size)
}